053-configurable-event-store/proposal.md

```diff
@@ -0,0 +1,450 @@
+`TRIGGER`s for creating/deleting `pipeline_build_events_x` and
+`team_build_events_x` tables when pipelines and teams are created/deleted.
+
+I propose that we comment out these up migrations (so future deployments don't
+get these tables/triggers), and leave any database initialization (or whatever
+appropriate backend setup) to `BuildEvent.Setup()`. This means `Setup` will
+have to be "smart" (i.e. have to handle the cases where the initialization has
+already been applied). It also makes it difficult to make changes to the
+schema, unlike with migrations - each `EventStore` (that stores data in a
+structured way - this may just end up being Postgres) will need to basically
+implement its own form of migrations if we need to change the schema. My guess
+is that we won't be changing the schema too often, though (given that we
+haven't changed `build_events` since the initial migration).
+
+
+## Configuration
+
+One thing I haven't really explored is how operators would configure the
+backend store. I think it could be similar to how Credential managers are
+configured - there are groups of independent configuration options for each
+possible `EventStore`. If certain flags are configured for a given backend, use
+that backend.
+
+
+# Open Questions
+
+* Does `Key` need to be `interface{}`, or can it be more specific like `uint`?
+  For backends that don't support auto-incrementing, but still need some
+  ordering (if the timestamp isn't sufficient), we **could** create a Postgres
+  sequence for every backend (external to the `EventStore`), and change the
+  signature to `Put(build db.Build, eventID uint, event atc.Event)`
+
+* Do we care about the `Last-Event-ID` header on the
+  `/api/v1/builds/{build_id}/events` endpoint? I don't think we use it anywhere
+  internally, but I guess we may want to keep it for backwards compatibility?
+  It's not exposed by `fly` or `go-concourse`, either.
+
+* Would any event stores benefit from bringing their own notification bus
+  implementation, rather than relying on Postgres' notification bus? Does using
+  the Postgres notification bus defeat the purpose of having an external store in
+  any way (e.g. could this be a bottleneck)?
+
+* Users won't be able to access build events that exist in the main Postgres
+  database if they switch to a new backend store. How can we make the
+  transition easier? One idea I have is to have a `FallbackEventStore` that
+  composes two `EventStores` - it will always `Put` to the primary one, but if
+  `Get` fails or returns no events from the primary `EventStore`, try the
+  fallback `EventStore`.
+
+
+# Answered Questions
+
+
+# New Implications
+
+* By default, there will be no change - Concourse will default to using a
+  Postgres `EventStore` using the same main Postgres database.
+* Operators will be able to configure a different store
+* If they configure a new store, they won't be able to access any existing
```

Comment:
I think this feature has to provide a migration solution, otherwise production clusters would be hard to move on external storage.

Reply:
Yeah I agree. I'd love to hear any thoughts you have on the best way to approach this. I can think of a couple options:

1. Introduce a `concourse migrate-build-events` command that migrates everything over to the configured build event store. It could possibly default to migrating from the Postgres database to whatever event store is configured, but it could also take a separate `from` configuration in case you want to migrate from a non-Postgres backend to a new backend.

2. Migrate out of Postgres "on the fly". i.e. we could have a component like the build reaper that migrates old builds to the new store, a few at a time. In this case, we'd probably still want some way to view build logs from Postgres in case they haven't yet been migrated, for which we could construct a `FallbackEventStore` that does something like this:

```go
type FallbackEventStore struct {
    EventStore
    Fallback EventStore
}

func (f *FallbackEventStore) Get(build db.Build, requested int, cursor *Key) ([]event.Envelope, error) {
    // first, try the main event store:
    events, err := f.EventStore.Get(build, requested, cursor)
    if err == nil && len(events) > 0 {
        return events, nil
    }
    // the build wasn't found (or at least has no events in the main store) - so try the Fallback (Postgres)
    // if this also errors, we'll probably want to return the err from the primary EventStore, but you get the idea
    return f.Fallback.Get(build, requested, cursor)
}
```

Reply:
I really like the option 2 having FallbackEventStore composing main and old eventstore. To add an extension to it, the fallback eventstore can also perform the migration(at background) when it read from an old eventstore.

```go
func (f *FallbackEventStore) Get(ctx context.Context, build db.Build, requested int, cursor *Key) ([]event.Envelope, error) {
   
  // first, try the main event store:
  ...

   events, err = f.Fallback.Get(build, requested, cursor)

   if err == nil && len(events) > 0 {
        //perhaps run this in the background to migrate this events to main eventstore
        f.EventStore.Put(ctx, build, events)

        return events, nil
    }
    return nil, err
}
```

The build event logs will be migrated only when a user access it for the first time. In most cases, users might not be interested looking at historical build events, so the expensive migration of all historical build event logs to main eventstore can be skipped and will be reaped after its retention period.

Reply:
I really like the option 2 having FallbackEventStore composing main and old eventstore. To add an extension to it, say lazy migration 😄 , the fallback eventstore can perform the migration(at background) when it read the events from an old eventstore.

```go
func (f *FallbackEventStore) Get(ctx context.Context, build db.Build, requested int, cursor *Key) ([]event.Envelope, error) {
   
  // first, try the main event store:
  ...

   events, err = f.Fallback.Get(build, requested, cursor)

   if err == nil && len(events) > 0 {
        //perhaps run this in the background to migrate this events to main eventstore
        f.EventStore.Put(ctx, build, events)

        return events, nil
    }
    return nil, err
}
```

The build event logs will be migrated only when a user access it for the first time. In most cases, users might not be interested looking at historical build events, so the expensive migration of all historical build event logs to main eventstore can be skipped and will be reaped after its retention period.

Reply:
Maybe just not migrate. Instead, add a flag to table `build` to indicate where is build logs of the build, PG or ES. So that reads old build logs from PG and new build logs from ES.

Reply:
Just a suggestion for another way to do the migration, I wonder if there could be a way to migrate over all the build events to their new event store without stopping the database? This will make it so that operators wanting to move over to a new build event store does not need to wait for the migrations to finish before starting concourse.

I imagine something like the operator would continue to have a running concourse deployment and the build events would currently be fetched from the postgres database. While the deployment is running, some tool can be used to migrate all the build events within the postgres database (old build event store) into the new build event store. Then once it has finished, the operator can redeploy concourse and point it at the new build event store. There will need to be some method for discarding the old data (maybe concourse can do that at startup if you configure a new event store? or maybe it can be part of the tool to wipe out build events?) 

I wonder for operators that want to configure a new build event store, they might want to benefit from being able to query or search for things within the build events. If we were to do the method of migrating on the fly, that would mean only certain build events would be migrated over. So taking that into account and the fact that operators that configure a new build event store most likely would like all their builds to be in this new store, I might be inclined to thinking that migrating all the data together would be preferable? I could totally be wrong about this though.

Reply:
@clarafu that definitely makes sense - having a separate tool that just points to the running Postgres is very much in line with https://github.com/concourse/rfcs/pull/53#discussion_r424022341.

FWIW the idea behind the "on the fly" approach is that it'd eventually migrate all build events over (not just for new events) - it'd essentially be doing what the tool you're suggesting is doing, but it's started by the ATC. I think it's beneficial to keep it separate though, for the reasons mentioned by @jchesterpivotal.

Reply:
Ohh I see! Sorry I missed where you clarified it by saying it will `build reaper that migrates old builds to the new store`. And yeah that is a good call that allowing the operator to migrate before upgrading is definitely a good idea.